test/widgets/lightbox_test.dart

```diff
@@ -197,6 +204,128 @@ class FakeVideoPlayerPlatform extends Fake
+      connection = store.connection as FakeApiConnection;
+      await store.addUser(eg.selfUser);
+
+      connection.prepare(json:
+        eg.newestGetMessagesResult(foundOldest: true, messages: [message]).toJson());
+
+      await tester.pumpWidget(TestZulipApp(accountId: eg.selfAccount.id,
+        child: MessageListPage(initNarrow: const CombinedFeedNarrow())));
+
+      await tester.pumpAndSettle();
+    }
+
+    testWidgets('Hero animation occurs smoothly when opening lightbox from message list', (tester) async {
+      prepareBoringImageHttpClient();
+
+      await setupMessageListPage(tester);
+
+      final messageContentFinder = find.byWidgetPredicate((widget) =>
+        widget is MessageContent && widget.message.id == message.id
+      );
+      final messageListImageFinder = find.descendant(
+        of: messageContentFinder,
+        matching: find.byType(RealmContentNetworkImage)
+      );
+      final initialImagePosition = tester.getRect(messageListImageFinder);
+      await tester.tap(messageListImageFinder);
+      await tester.pump();
+      // pump to start hero animation
+      await tester.pump();
+
+      final heroAnimationDuration = Duration(milliseconds: 300);
+      final steps = 150;
+      final stepDuration = heroAnimationDuration ~/ steps;
+      List<Rect> animatedPositions = [];
+      for (int i = 1; i <= steps; i++) {
+        await tester.pump(stepDuration);
+
+        final animatedFlightImageFinder = find.byWidgetPredicate((widget) =>
+          widget is RealmContentNetworkImage && widget.src == imageSrcUrl
+        );
+        animatedPositions.add(tester.getRect(animatedFlightImageFinder));
+      }
+
+      final totalDistance = sqrt(
+        pow(initialImagePosition.top - animatedPositions.last.top, 2) +
+        pow(initialImagePosition.left - animatedPositions.last.left, 2));
+
+      Rect previousPosition = initialImagePosition;
+      double maxStepDistance = 0.0;
+      for (final position in animatedPositions) {
+        final stepDistance = sqrt(
+          pow(position.top - previousPosition.top, 2) +
+          pow(position.left - previousPosition.left, 2));
+        maxStepDistance = max(maxStepDistance, stepDistance);
+        check(position).not((pos) => pos.equals(previousPosition));
+
+        previousPosition = position;
+      }
+      check(maxStepDistance).isLessThan(0.03 * totalDistance);
+
+      debugNetworkImageHttpClientProvider = null;
+    });
+
+    testWidgets('no hero animation occurs between different message list pages for same image', (tester) async {
+      prepareBoringImageHttpClient();
+
+      await setupMessageListPage(tester);
+
+      final imageFinder = find.byWidgetPredicate((widget) =>
+        widget is RealmContentNetworkImage && widget.src == imageSrcUrl
+      );
+      final firstImagePosition = tester.getRect(imageFinder);
+      final firstElement = imageFinder.evaluate().single;
```

Comment:
Bit cleaner (as `evaluate` is meant really as internal API for implementing these finders):
```suggestion
      final firstElement = tester.element(imageFinder);
```